codes/models/modules/DUF_arch.py

```diff
@@ -0,0 +1,682 @@
+
+    def forward(self, x):
+        for i in range(0, len(self.dense_blocks), 6):
+            y = x.clone()
+            for j in range(6):
+                y = self.dense_blocks[i + j](y)
+            x = torch.cat((x, y), 1)
+        return x
+
+
+class DUF_52L(nn.Module):
+    ''' same structure as that in the paper
+    https://github.com/yhjo09/VSR-DUF'''
+
+    def __init__(self, adapt_official=False):
+        super(DUF_52L, self).__init__()
+        self.conv3d_1 = nn.Conv3d(3, 64, (1, 3, 3), stride=(1, 1, 1), padding=(0, 1, 1), bias=True)
+        self.dense_block_1 = DenseBlock_52L_A(64, 16)  # 64 + 21 * 9 = 400, T = 7
+        self.dense_block_2 = DenseBlock(400, 16, t_reduce=True)  # 400 + 16 * 3 = 448, T = 1
+
+        self.bn3d_2 = nn.BatchNorm3d(448, eps=1e-3, momentum=1e-3)
+        self.conv3d_2 = nn.Conv3d(448, 256, (1, 3, 3), stride=(1, 1, 1), padding=(0, 1, 1),
+                                  bias=True)
+
+        self.conv3d_r1 = nn.Conv3d(256, 256, (1, 1, 1), stride=(1, 1, 1), padding=(0, 0, 0),
+                                   bias=True)
+        self.conv3d_r2 = nn.Conv3d(256, 3 * 16, (1, 1, 1), stride=(1, 1, 1), padding=(0, 0, 0),
+                                   bias=True)
+
+        self.conv3d_f1 = nn.Conv3d(256, 512, (1, 1, 1), stride=(1, 1, 1), padding=(0, 0, 0),
+                                   bias=True)
+        self.conv3d_f2 = nn.Conv3d(512, 1 * 5 * 5 * 16, (1, 1, 1), stride=(1, 1, 1),
+                                   padding=(0, 0, 0), bias=True)
+
+        self.dynamic_filter = DynamicUpsamplingFilter_3C((1, 5, 5))
+
+        # initialization
+        _initialize_weights(self.conv3d_1, 0.1)
+        _initialize_weights(self.bn3d_2, 0.1)
+        _initialize_weights(self.conv3d_2, 0.1)
+        _initialize_weights(self.conv3d_r1, 0.1)
+        _initialize_weights(self.conv3d_r2, 0.1)
+        _initialize_weights(self.conv3d_f1, 0.1)
+        _initialize_weights(self.conv3d_f2, 0.1)
+
+        self.adapt_official = adapt_official
+
+    def forward(self, x):
+        '''
+        x: [B, T, C, H, W], T = 7. reshape to [B, C, T, H, W] for Conv3D
+        Generate filters and image residual:
+        Fx: [B, 25, 16, H, W] for DynamicUpsamplingFilter_3C
+        Rx: [B, 3*16, 1, H, W]
+        '''
+        B, T, C, H, W = x.size()
+        x = x.permute(0, 2, 1, 3, 4)  # [B,C,T,H,W] for Conv3D
+        x_center = x[:, :, T // 2, :, :]
+        x = self.conv3d_1(x)
+        x = self.dense_block_1(x)
+        x = self.dense_block_2(x)
+        x = F.relu(self.conv3d_2(F.relu(self.bn3d_2(x), inplace=True)), inplace=True)
+
+        # image residual
+        Rx = self.conv3d_r2(F.relu(self.conv3d_r1(x), inplace=True))  # [B, 3*16, 1, H, W]
+
+        # filter
+        Fx = self.conv3d_f2(F.relu(self.conv3d_f1(x), inplace=True))  # [B, 25*16, 1, H, W]
+        Fx = F.softmax(Fx.view(B, 25, 16, H, W), dim=1)
+
+        # Adapt to official model weights
+        if self.adapt_official:
```

Comment:
This adapt_official part has appeared several times. Can it be separated into a function?